src/batcher/solve_pair.rs

```diff
@@ -1,16 +1,438 @@
+    /// and vice versa.
+    fn opposite_tokens(&self, other: &Order) -> bool;
+    /// Returns whether there is a price that satisfies both orders.
+    fn have_price_overlap(&self, other: &Order) -> bool;
+}
+
+
+impl Matchable for Order {
+    fn match_compare(&self, other: &Order) -> Option<OrderPairType> {
+        if !self.attracts(other) {
+            return None;
+        }
+
+        if self.buy_amount <= other.sell_amount && self.sell_amount <= other.buy_amount {
+            Some(OrderPairType::LhsFullyFilled)
+        } else if self.buy_amount >= other.sell_amount && self.sell_amount >= other.buy_amount {
+            Some(OrderPairType::RhsFullyFilled)
+        } else {
+            Some(OrderPairType::BothFullyFilled)
+        }
+    }
+
+    fn attracts(&self, other: &Order) -> bool {
+        self.opposite_tokens(other) && self.have_price_overlap(other)
+    }
+
+    fn opposite_tokens(&self, other: &Order) -> bool {
+        self.buy_token == other.sell_token && self.sell_token == other.buy_token
+    }
+
+    fn have_price_overlap(&self, other: &Order) -> bool {
+        self.sell_amount > U256::zero()
+            && other.sell_amount > U256::zero()
+            && self.buy_amount * other.buy_amount <= self.sell_amount * other.sell_amount
+    }
+}
+
+struct Match {
+    order_pair_type: OrderPairType,
+    orders: OrderPair,
+}
+
+type OrderPair = [Order; 2];
+
+type MatchedAmounts = [U256; 2];
+
 
 pub fn solve_pair(
     sell_orders_token0: &Vec<Order>,
     sell_orders_token1: &Vec<Order>,
 ) -> Result<Solution> {
+    // Can we assume that the orders are sorted already?
+    assert!(check_orders_sorted_by_limit_price(&sell_orders_token0));
+    assert!(check_orders_sorted_by_limit_price(&sell_orders_token1));
+    // Or do we need to sort them first here?
+    //sell_orders_token0.sort();
+    //sell_orders_token1.sort();
+
+    // Get number of orders in each direction.
+    let nr_orders_token0 = sell_orders_token0.len();
+    let nr_orders_token1 = sell_orders_token1.len();
+
+    // Init vectors of sell amounts.
+    let mut sell_volumes_token0 = vec![U256::zero(); nr_orders_token0];
+    let mut sell_volumes_token1 = vec![U256::zero(); nr_orders_token1];
+
+    // Match orders with best limit prices, if possible.
+    if !(sell_orders_token0.is_empty() || sell_orders_token1.is_empty()) {
+
+        // The best orders are the last elements in their vectors.
+        let best_sell_order_token0 = &sell_orders_token0[nr_orders_token0 - 1];
+        let best_sell_order_token1 = &sell_orders_token1[nr_orders_token1 - 1];
+
+        if let Some(order_pair_type) = best_sell_order_token0.match_compare(
+            &best_sell_order_token1
+        ) {
+            let matched_amounts = get_matched_amounts(
+                &Match {
+                    order_pair_type,
+                    orders: [best_sell_order_token0.clone(), best_sell_order_token1.clone()],
```

Comment:
Right - I pushed an update doesn't require the `clone()` (but other weird things I don't really understand :see_no_evil:)

Reply:
The original version was borrowed again from https://github.com/gnosis/dex-services/blob/56d192b558f0e66e6c18c269e5bd1549dd990f20/services-core/src/price_finding/naive_solver.rs#L166-L169

Reply:
Wow! I'm sorry I asked, it looks like the answer is even more intriguing. I'll see if I can find out what's going on ;)

Reply:
> Just curious: why does rust need clone here?

This is because of Rust ownership semantics. In layman's terms, only one binding (aka variable) can "own" any value at any time. If the `clone()` were removed, then the value would be "moved", i.e. the ownership would have been transferred from `sell_orders_token0` vec to the `Match {...}` instance. This is not allowed here, since `sell_orders_token0` is itself a "borrowed" value, so "moving" out of it is prohibited.

I would say that this is the concept with the steepest learning curve in Rust. I suggest looking at the chapter in the book about ownership and borrowing which probably explains it much better than I do:
https://doc.rust-lang.org/book/ch04-00-understanding-ownership.html

Reply:
Thanks Nick! I was reading about it yesterday. It's (almost) clear now ;)